### PR TITLE
Ability to use all fast forward modes available

### PR DESCRIPTION
**Pull request by martinda**
I would like to help provide the feature requested in https://issues.jenkins-ci.org/browse/JENKINS-12402 as I need the same feature. I am starting by editing the git-client-plugin and I would like a review please. Specifically, I added a @Bug(12402) to the first new test I added, but is that appropriate? Also, I have not found where to add a test for the CliGitApiImpl, I only found where to add tests for the JGit implementation. Your feedback and comments are welcome.

### CODE REVIEW

**Comment by MarkEWaite** [comment]
I fetched this proposal into my git client plugin test setup and confirmed that the tests you added provided the code coverage I expected, and the tests executed successfully on various versions of Windows (7, 8, Home Server 2011) with at least one of those being 32 bit and one a 64 bit version.  The tests also ran successfully on Debian 6 x86 "squeeze", Debian 7 x64 "wheezy", Debian 8 x64 "testing", CentOS 6 x86, and CentOS 7 x64.

That combination of tests also confirms that the tests behave as expected with several different git versions ranging from 1.7.2 to 2.2.1.

**Comment by martinda** [comment]
I learned a lot from your review. Let me know what's next, I am happy to contribute. I am working on the UI changes in the mean time. Thank you.

**Comment by MarkEWaite** [comment]
I added one more comment to your change.  Once you've made that change,
I'll evaluate the pull request again through my environment and merge or
rebase it into the plugin.

On Thu, Jan 29, 2015 at 7:30 PM, martinda <notifications@github.com> wrote:

> I learned a lot from your review. Let me know what's next, I am happy to
> contribute. I am working on the UI changes in the mean time. Thank you.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/git-client-plugin/pull/159#issuecomment-72144146>
> .
>



-- 
Thanks!
Mark Waite

**Comment by martinda** [comment]
Regarding the UI changes in the git-plugin, should I be changing the UserMergeOptions() constructor and add a new parameter, or should I add a new constructor with the new fastForward mode parameter and have two constructors?
Also, the two plugins need to be kept in sync, and I am not sure how to do that (for now I changed my git pom to point to 1.15.1-SNAPSHOT of git-client to get going).
Thanks again.

**Comment by martinda** [comment]
I have completed the changes to the UI, which can be seen here:
https://github.com/martinda/git-plugin/compare/feature/JENKINS-12402-add-option-for-no-ff-when-merging#diff-1c4d62110a96db9b472b7a8c4a0e14e2
However I do not know if I should be creating a pull request for the UI now, or if I should wait until this PR has been pulled in and released.

**Comment by MarkEWaite** [comment]
Wait for the git-client-plugin pull request to complete and a version of
the git-client-plugin to release.  That will allow you to submit a pull
request based on the released version of the git-client-plugin with the
changes you need in the git-plugin.

Likely I'll release a new version of the git-client-plugin with your
changes at the end of this week.

Mark Waite

On Mon, Feb 2, 2015 at 12:06 PM, martinda <notifications@github.com> wrote:

> I have completed the changes to the UI, which can be seen here:
>
> https://github.com/martinda/git-plugin/compare/feature/JENKINS-12402-add-option-for-no-ff-when-merging#diff-1c4d62110a96db9b472b7a8c4a0e14e2
> However I do not know if I should be creating a pull request for the UI
> now, or if I should wait until this PR has been pulled in and released.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/git-client-plugin/pull/159#issuecomment-72516746>
> .
>



-- 
Thanks!
Mark Waite

**Comment by KostyaSha** [comment]
:+1: Just started searching --no-ff

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java [hunk 1]
```diff
@@ -2168,6 +2168,122 @@ public void test_merge_strategy_correct_fail() throws Exception {
         }
     }
 
+    @Bug(12402)
```
Comment (KostyaSha) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: This is deprecated, use `@Issue`

**Comment by KostyaSha** [comment]
Please open PR to git-plugin and mark in title DON"T MERGE with first comment that PR pending git-client-plugin release. I will start review process.

**Comment by KostyaSha** [comment]
Sorry, i opened PR myself from your repo/branch https://github.com/jenkinsci/git-plugin/pull/299

File: src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java (hunk 1 above)
Comment (martinda) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: Thanks for letting me know. I was just copying what I saw elsewhere in that file. Would you like me to change all the @Bug in this project or just this one? Looks like a clean up activity that should be on its own PR isn't it?

Comment (KostyaSha) [inline src/test/java/org/jenkinsci/plugins/gitclient/GitAPITestCase.java]: It not critical i think @MarkEWaite will change it before merge. Don't copy what you see without understanding.` IDE shows strikethrough text, then look at description that describes why it deprecated.
